File: examples/ruby/use_the_connection.rb
# Step 3 — Use the connection (Ruby / Rails)
#
# One helper covers every app and action. Reads and writes are the same call
# with a different action slug.
class MindCloud
  API = "https://connect.mindcloud.co".freeze

  # Action slugs are the docs-style kebab form: listChannels -> list-channels.
  def self.run_action(app_slug:, action_slug:, installation_id:, arguments: {})
    response = HTTP.headers(headers).post(
      "#{API}/v2/universal/apps/#{app_slug}/actions/#{action_slug}/run",
      json: { installationId: installation_id, arguments: arguments }
    )

    response.parse
  end

  # Read: the customer's Slack channels.
  def self.slack_channels(installation_id)
    run_action(
      app_slug: "slack",
      action_slug: "list-channels",
      installation_id: installation_id,
      arguments: { excludeArchived: true }
    ).fetch("data").map { |row| row.slice("id", "name") }
  end

  # Write: post as that customer.
  def self.send_slack_message(installation_id, channel_id, text)
    run_action(
      app_slug: "slack",
      action_slug: "send-channel-message",
      installation_id: installation_id,
      arguments: { channel: channel_id, text: text }
    )
  end

  def self.headers
    {
      "Authorization" => "Bearer #{ENV.fetch('MINDCLOUD_API_KEY')}",
      "Content-Type" => "application/json"
    }
  end
end

File: examples/node/use-the-connection.js
// Step 3 — Use the connection (Node.js / Express)
//
// One helper covers every app and action. Reads and writes are the same call
// with a different action slug.
const API = 'https://connect.mindcloud.co';

const headers = {
  Authorization: `Bearer ${process.env.MINDCLOUD_API_KEY}`,
  'Content-Type': 'application/json'
};

// Action slugs are the docs-style kebab form: listChannels → list-channels.
const runAction = async ({ appSlug, actionSlug, installationId, args }) => {
  const response = await fetch(`${API}/v2/universal/apps/${appSlug}/actions/${actionSlug}/run`, {
    method: 'POST',
    headers,
    body: JSON.stringify({ installationId, arguments: args })
  });

  return response.json();
};

// Read: the customer's Slack channels.
router.get('/api/slack/channels', async (req, res) => {
  const installationId = await installationForUser(req.session.user);

  const body = await runAction({
    appSlug: 'slack',
    actionSlug: 'list-channels',
    installationId,
    args: { excludeArchived: true }
  });

  res.json(body.data.map(({ id, name }) => ({ id, name })));
});

// Write: post as that customer.
router.post('/api/slack/send', async (req, res) => {
  const installationId = await installationForUser(req.session.user);

  const body = await runAction({
    appSlug: 'slack',
    actionSlug: 'send-channel-message',
    installationId,
    args: { channel: req.body.channelId, text: req.body.text }
  });

  res.json(body);
});
